File: apps/ios/Burcum/Services/AuthManager.swift
import Foundation
import SwiftUI

struct User: Codable, Identifiable {
    let id: String
    let email: String
    let name: String
    let birthDate: String
    let birthTime: String?
    let birthCity: String?
    let zodiacSign: String
    let emailVerifiedAt: String?
    let subscriptionTier: SubscriptionTier
    let dailyReadingsCount: Int?
    let lastReadingDate: String?
    let createdAt: String?

    var emailVerified: Bool {
        emailVerifiedAt != nil
    }
}

enum SubscriptionTier: String, Codable {
    case free
    case premium
    case vip
}

@MainActor
class AuthManager: ObservableObject {
    @Published var user: User?
    @Published var isAuthenticated = false
    @Published var isLoading = false
    @Published var error: String?

    private let apiClient = APIClient.shared
    private let keychain = KeychainManager.shared

    init() {
        if let savedUser = keychain.getUser() {
            self.user = savedUser
            self.isAuthenticated = true
            Task {
                await refreshUser()
            }
        }
    }

    var isPremium: Bool {
        user?.subscriptionTier == .premium || user?.subscriptionTier == .vip
    }

    var isVIP: Bool {
        user?.subscriptionTier == .vip
    }

    func login(email: String, password: String) async {
        isLoading = true
        error = nil

        do {
            let response = try await apiClient.login(email: email, password: password)
            guard response.success, let user = response.data else {
                self.error = response.error ?? "Giris basarisiz"
                isLoading = false
                return
            }
            self.user = user
            self.isAuthenticated = true
            _ = keychain.saveUser(user)
        } catch let apiError as APIError {
            self.error = apiError.errorDescription
        } catch {
            self.error = "Baglanti hatasi. Lutfen tekrar deneyin."
        }

        isLoading = false
    }

    func register(email: String, password: String, name: String?, birthDate: Date?) async {
        isLoading = true
        error = nil

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"

        let request = RegisterRequest(
            email: email,
            password: password,
            name: name ?? "",
            birthDate: birthDate.map { formatter.string(from: $0) } ?? "",
            birthTime: nil,
            birthCity: nil
        )

        do {
            let response = try await apiClient.register(request: request)
            guard response.success, let user = response.data else {
                self.error = response.error ?? "Kayit basarisiz"
                isLoading = false
                return
            }
            self.user = user
            self.isAuthenticated = true
            _ = keychain.saveUser(user)
        } catch let apiError as APIError {
            self.error = apiError.errorDescription
        } catch {
            self.error = "Baglanti hatasi. Lutfen tekrar deneyin."
        }

        isLoading = false
    }

    func logout() {
        Task {
            do {
                let _: APIResponse<LogoutResponse> = try await apiClient.logout()
            } catch {
                // Silent fail
            }
        }
        clearLocalSession()
    }

    func deleteAccount() async -> Bool {
        isLoading = true
        error = nil

        do {
            let response = try await apiClient.deleteAccount()
            guard response.success else {
                self.error = response.error ?? "Hesap silme basarisiz"
                isLoading = false
                return false
            }

            clearLocalSession()
            isLoading = false
            return true
        } catch let apiError as APIError {
            self.error = apiError.errorDescription
        } catch {
            self.error = "Hesap silinirken bir hata olustu. Lutfen tekrar deneyin."
        }

        isLoading = false
        return false
    }

    func refreshUser() async {
        do {
            let response = try await apiClient.getCurrentUser()
            if response.success, let updatedUser = response.data {
                self.user = updatedUser
                self.isAuthenticated = true
                _ = keychain.saveUser(updatedUser)
            } else if !response.success {
                clearLocalSession()
            }
        } catch let apiError as APIError {
            if case .unauthorized = apiError {
                clearLocalSession()
            }
        } catch {
            // Network error: keep cached user
        }
    }

    private func clearLocalSession() {
        apiClient.clearCookies()
        user = nil
        isAuthenticated = false
        keychain.clearAll()
    }
}

File: apps/android/app/src/main/java/com/burcum/app/AuthManager.kt
package com.burcum.app

import android.content.Context
import com.burcum.app.model.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class AuthManager(context: Context) {
    private val secureStorage = SecureStorage(context)
    private val apiClient = BurcumApp.instance.apiClient

    private val _user = MutableStateFlow<User?>(null)
    val user: StateFlow<User?> = _user.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    val isAuthenticated: Boolean
        get() = _user.value != null

    val isPremium: Boolean
        get() = _user.value?.subscriptionTier == SubscriptionTier.PREMIUM ||
                _user.value?.subscriptionTier == SubscriptionTier.VIP

    val isVIP: Boolean
        get() = _user.value?.subscriptionTier == SubscriptionTier.VIP

    init {
        // Load saved user
        secureStorage.getUser()?.let { savedUser ->
            _user.value = savedUser
        }
    }

    suspend fun login(email: String, password: String): Boolean {
        _isLoading.value = true
        _error.value = null

        return try {
            val response = apiClient.api.login(LoginRequest(email, password))
            if (response.success && response.data != null) {
                secureStorage.saveToken(response.data.token)
                secureStorage.saveUser(response.data.user)
                _user.value = response.data.user
                true
            } else {
                _error.value = response.error ?: "Giriş başarısız"
                false
            }
        } catch (e: Exception) {
            _error.value = e.message ?: "Bağlantı hatası"
            false
        } finally {
            _isLoading.value = false
        }
    }

    suspend fun register(email: String, password: String, name: String?, birthDate: String?): Boolean {
        _isLoading.value = true
        _error.value = null

        return try {
            val response = apiClient.api.register(
                RegisterRequest(email, password, name, birthDate)
            )
            if (response.success && response.data != null) {
                secureStorage.saveToken(response.data.token)
                secureStorage.saveUser(response.data.user)
                _user.value = response.data.user
                true
            } else {
                _error.value = response.error ?: "Kayıt başarısız"
                false
            }
        } catch (e: Exception) {
            _error.value = e.message ?: "Bağlantı hatası"
            false
        } finally {
            _isLoading.value = false
        }
    }

    suspend fun logout() {
        try {
            apiClient.api.logout()
        } catch (e: Exception) {
            // Ignore logout errors
        }

        secureStorage.clearAll()
        _user.value = null
    }

    suspend fun refreshUser() {
        try {
            val response = apiClient.api.getCurrentUser()
            if (response.success && response.data != null) {
                secureStorage.saveUser(response.data)
                _user.value = response.data
            }
        } catch (e: Exception) {
            // Silent fail
        }
    }

    fun clearError() {
        _error.value = null
    }
}
